weight_round_robin/weight_round_robin.go

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2022 CloudWeGo Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package weightroundrobin
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/cloudwego/hertz/pkg/app/client/discovery"
+	"github.com/cloudwego/hertz/pkg/app/client/loadbalance"
+	"github.com/cloudwego/hertz/pkg/common/hlog"
+	"golang.org/x/sync/singleflight"
+)
+
+type weightRoundRobinBalancer struct {
+	cachedInfo sync.Map
+	sfg        singleflight.Group
+}
+
+type weightRoundRobinInfo struct {
+	instances       []discovery.Instance
+	effectiveWeight []int32
+	currentWeight   []int32
+	weightSum       int
+}
+
+// NewWeightRoundRobinBalancer creates a loadbalancer using round-robin algorithm.
+func NewWeightRoundRobinBalancer() loadbalance.Loadbalancer {
+	lb := &weightRoundRobinBalancer{}
+	return lb
+}
+
+func (rr *weightRoundRobinBalancer) calcWeightInfo(e discovery.Result) *weightRoundRobinInfo {
+	w := &weightRoundRobinInfo{
+		instances:       make([]discovery.Instance, len(e.Instances)),
+		effectiveWeight: make([]int32, len(e.Instances)),
+		currentWeight:   make([]int32, len(e.Instances)),
+		weightSum:       0,
+	}
+
+	var cnt int
+
+	for idx := range e.Instances {
+		weight := e.Instances[idx].Weight()
+		if weight > 0 {
+			w.instances[cnt] = e.Instances[idx]
+			w.effectiveWeight[cnt] = int32(weight)
+			w.currentWeight[cnt] = 0
+			w.weightSum += weight
+			cnt++
+		} else {
+			hlog.SystemLogger().Warnf("Invalid weight=%d on instance address=%s", weight, e.Instances[idx].Address())
+		}
+	}
+
+	w.instances = w.instances[:cnt]
+	return w
+}
+
+// Pick implements the Loadbalancer interface.
+func (rr *weightRoundRobinBalancer) Pick(e discovery.Result) discovery.Instance {
+	ri, ok := rr.cachedInfo.Load(e.CacheKey)
+	if !ok {
```

Comment:
Also, if here can be false? If it is false, maybe there is no instance? I don't review when will call Rebalance

Reply:
> 

When it is false, we can't use e.CacheKey to get weightRoundRobinInfo from sync.Map, so we will enter it and call the Do method to store the return value of the calcWeightInfo method into the Map. It's not that there are no instance.